File: static/js/admin/product.js
var product_sidebar_add = document.getElementById('product_sidebar_add');
var product_sidebar_remove = document.getElementById('product_sidebar_remove');
var product_sidebar_update = document.getElementById('product_sidebar_update');

product_sidebar_add.onclick = function() {
    window.location.href = 'http://127.0.0.1:5000/admin/product/add';
}

product_sidebar_remove.onclick = function() {
    window.location.href = 'http://127.0.0.1:5000/admin/product/remove';
}

product_sidebar_update.onclick = function() {
    window.location.href = 'http://127.0.0.1:5000/admin/product/update';
}


var btnAddProduct = document.getElementById('btnAddProduct');
if (btnAddProduct){
    btnAddProduct.onclick = function(){
        fetch("http://127.0.0.1:5000/admin/product/add", {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify({
                name: document.getElementById('name').value, //Điện thoại OPPO Reno10 5G (8GB/256GB) - Hàng Chính Hãng
                brand_name: document.getElementById('brand_name').value,
                price: document.getElementById('price').value,
                original_price: document.getElementById('original_price').value,
                thumbnail_url: document.getElementById('thumbnail_url').value,
                category: document.getElementById('category').value,
            })
        })
        .then(response => response.json())
        .then(data => {
            console.log(data);
            alert('Product added successfully');
        })
        .catch(error => {console.error('GET request error:', error)});
    }
}


var search_input = document.getElementById('search_input');
var optionList = document.getElementById('option');
var showList = document.getElementById('showList');

if (search_input){
    search_input.addEventListener('keyup', function(){
        fetch("http://127.0.0.1:5000/admin/product/remove/getSuggestion", {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify({
                'query': search_input.value,
                'option': optionList.value,
            })
        })
        .then(response => response.json())
        .then(data => { 
            showList.innerHTML = '';

            for (var i = 0; i < data['data'].length; i++){
                var option = document.createElement('option');
                option.value = data['data'][i][optionList.value];
                option.textContent = data['data'][i][optionList.value];
                showList.appendChild(option);
            }
        })
        .catch(error => {console.error('GET request error:', error)});
    });
}

var searchRemove = document.getElementById('searchRemove');
var searchUpdate = document.getElementById('searchUpdate');
var showDetailProduct = document.getElementById('showDetailProduct');

if (searchRemove){
    searchRemove.onclick = function() {
        fetch("http://127.0.0.1:5000/admin/product/remove/getProduct", {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify({
                'option': optionList.value,
                'dataProduct': showList.value,
            })
        })
        .then(response => response.json())
        .then(data => {
            console.log(data);
            showDetailProduct.innerHTML = '';
            var data = data.data;
    
            var table = document.createElement("table");
            table.style.width = "90%";
            table.style.margin = "20px";
    
            function addTrTag(labelData, valueData){
                var tr = document.createElement("tr");
            
                var label = document.createElement("td");
                label.style.width = '200px';
                label.style.paddingLeft = '10px';
                label.textContent = labelData;
                tr.appendChild(label);
            
                var value = document.createElement("td");
                value.style.paddingLeft = '10px';
                value.textContent = valueData;
                tr.appendChild(value);
            
                table.appendChild(tr);
            }
    
            addTrTag('Name', data['name']);
            addTrTag('ID', data['id']);
            addTrTag('Brand', data['brand_name']);
            addTrTag('Price', data['price']);
            addTrTag('Original Price', data['original_price']);
            addTrTag('Quantity', data['quantity_sold']);
            addTrTag('Rating', data['rating_average']);
            addTrTag('Category', data['category']);
            
    
            var btnRemove = document.createElement('button');
            btnRemove.id = 'btnRemoveProduct';
            btnRemove.textContent = 'Remove';
    
            btnRemove.onclick = function() {
                fetch("http://127.0.0.1:5000/admin/product/remove/removeProduct", {
                    method: 'POST',
                    headers: {'Content-Type': 'application/json'},
                    body: JSON.stringify({
                        'option': optionList.value,
                        'dataProduct': showList.value,
                    })
                })
                .then(response => response.json())
                .then(data => {
                    console.log(data);
                    alert(data['message']);
                })
                .catch(error => {console.error('GET request error:', error)});
            };
    
            showDetailProduct.appendChild(table);
            showDetailProduct.appendChild(btnRemove);
        })
        .catch(error => {console.error('GET request error:', error)});
    };
}

if (searchUpdate){
    searchUpdate.onclick = function() {
        fetch("http://127.0.0.1:5000/admin/product/update/getProduct", {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify({
                'option': optionList.value,
                'dataProduct': showList.value,
            })
        })
        .then(response => response.json())
        .then(data => {
            showDetailProduct.innerHTML = '';
            var data = data.data;
    
            var table = document.createElement("table");
            table.style.width = "90%";
            table.style.margin = "20px";
    
            function addTrTag(labelData, valueData = ''){
                var tr = document.createElement("tr");
            
                var label = document.createElement("td");
                label.style.width = '200px';
                label.style.paddingLeft = '10px';
                label.textContent = labelData;
                tr.appendChild(label);
            
                var value = document.createElement("td");
                var input = document.createElement("input");
                input.type = "text";
                input.className = 'textUpdate';
                input.value = valueData;
                value.append(input);
                tr.appendChild(value);
            
                table.appendChild(tr);
            }
    
            addTrTag('Name', data['name']);
            addTrTag('ID', data['id']);
            addTrTag('Brand', data['brand_name']);
            addTrTag('Price', data['price']);
            addTrTag('Original Price', data['original_price']);
            addTrTag('Category', data['category']);
            
            var btnUpdate = document.createElement('button');
            var textUpdate = document.getElementsByClassName('textUpdate');
            btnUpdate.id = 'btnUpdateProduct';
            btnUpdate.textContent = 'Update';
    
            btnUpdate.onclick = function() {
                var oldID = data['id'];
                var newID = parseInt(textUpdate[1].value);
                var newprice = parseInt(textUpdate[3].value);
                var newOriginalPrice = parseInt(textUpdate[4].value);
                var newCategory = parseInt(textUpdate[5].value);

                data['name'] = textUpdate[0].value;
                data['id'] = newID;
                data['brand_name'] = textUpdate[2].value;
                data['price'] = newprice;
                data['original_price'] = newOriginalPrice;
                data['category'] = newCategory;
                data['discount'] = newOriginalPrice - newprice;
                data['discount_rate'] = 100 - parseInt(newprice / newOriginalPrice * 100);

                fetch("http://127.0.0.1:5000/admin/product/update/updateProduct", {
                    method: 'POST',
                    headers: {'Content-Type': 'application/json'},
                    body: JSON.stringify({
                        'info': data,
                        'id': oldID
                    })
                })
                .then(response => response.json())
                .then(data => {
                    console.log(data);
                    alert(data['query']);
                })
                .catch(error => {console.error('GET request error:', error)});
            };
    
            showDetailProduct.appendChild(table);
            showDetailProduct.appendChild(btnUpdate);
        })
        .catch(error => {console.error('GET request error:', error)});
    };
}

File: admin/product.py
import random
from utils.database import connectDB
from admin.utils import convert_to_slug
from flask import Blueprint, request, jsonify, render_template


admin_product_bp = Blueprint('adminProduct', __name__)

db = connectDB()
productDB = db["Product"]

@admin_product_bp.route("/", methods=["GET"])
def product():
    return render_template("admin/product.html")


@admin_product_bp.route("/add", methods=["GET", "POST"])
def addProduct():
    if request.method == "GET":
        return render_template("admin/product.html", type="add")

    if request.method == "POST":
        data = request.get_json()
        name = data.get("name")
        brand_name = data.get("brand_name")
        price = int(data.get("price"))
        original_price = int(data.get("original_price"))
        thumbnail_url = data.get("thumbnail_url")
        category = int(data.get("category"))
        url_key = convert_to_slug(name) + "-p" + str(random.randint(1, 99999999))
        discount_rate = int(100 - (price / original_price * 100))
        id = random.randint(1, 99999999)

        data = {
            "id": id, "spid": id, "name": name, "url_key": url_key, "brand_name": brand_name,
            "price": price, "original_price": original_price, "discount": original_price - price,
            "discount_rate": discount_rate, "rating_average": 0,"review_count": 0, 
            "thumbnail_url": thumbnail_url, "quantity_sold": 0, "category": category
        }
        
        productDB.insert_one(data)
        return jsonify({"success": True})


@admin_product_bp.route("/remove", methods=["GET"])
def removeProduct():
    if request.method == "GET":
        return render_template("admin/product.html", type="remove")

    if request.method == "POST":
        data = request.get_json()
        name = data.get("name")
        return jsonify({"name": name})


@admin_product_bp.route("/remove/getSuggestion", methods=["POST"])
def get_suggestion_remove():
    data = request.get_json()
    option = data.get("option", "")
    query = data.get("query", "")

    if query != "" and option == "id":
        num_less = int(query + ("0" * (9 - len(query))))
        num_greater = int(query + ("9" * (9 - len(query))))

        result = productDB.find(
            {option: {"$type": "int", "$gte": num_less, "$lte": num_greater}},
            {"_id": 0, option: 1},
        )
        return jsonify({"data": list(result), "query": query, "option": option})

    elif query != "" and option == "name":
        result = productDB.find({option: {"$regex": query, "$options": "i"}}, {"_id": 0, option: 1})
        return jsonify({"data": list(result), "query": query, "option": option})

    else:
        return jsonify({"data": [], "query": query, "option": option})


@admin_product_bp.route("/remove/getProduct", methods=["POST"])
def get_product_remove():
    data = request.get_json()
    option = data.get("option", "")
    query = data.get("dataProduct", "")

    if option == "id":
        query = int(query)

    result = productDB.find_one({option: query}, {"_id": 0})
    return jsonify({"data": result, "query": query, "option": option})


@admin_product_bp.route("/remove/removeProduct", methods=["POST"])
def remove_product():
    data = request.get_json()
    option = data.get("option", "")
    query = data.get("dataProduct", "")

    if option == "id":
        query = int(query)

    result = productDB.delete_one({option: query})

    if result.deleted_count == 1:
        return jsonify({"query": query, "option": option, "message": "Removed successfully"})

    return jsonify({"query": query, "option": option, "message": "Removed failed"})


@admin_product_bp.route("/update", methods=["GET"])
def updateProduct():
    if request.method == "GET":
        return render_template("admin/product.html", type="update")

    if request.method == "POST":
        data = request.get_json()
        name = data.get("name")
        return jsonify({"name": name})


@admin_product_bp.route("/update/getSuggestion", methods=["POST"])
def get_suggestion_update():
    data = request.get_json()
    option = data.get("option", "")
    query = data.get("query", "")

    if query != "" and option == "id":
        num_less = int(query + ("0" * (9 - len(query))))
        num_greater = int(query + ("9" * (9 - len(query))))

        result = productDB.find(
            {option: {"$type": "int", "$gte": num_less, "$lte": num_greater}},
            {"_id": 0, option: 1},
        )

        return jsonify({"data": list(result), "query": query, "option": option})

    elif query != "" and option == "name":
        result = productDB.find(
            {option: {"$regex": query, "$options": "i"}}, 
            {"_id": 0, option: 1}
        )
        return jsonify({"data": list(result), "query": query, "option": option})

    else:
        return jsonify({"data": [], "query": query, "option": option})


@admin_product_bp.route("/update/getProduct", methods=["POST"])
def get_product_update():
    data = request.get_json()
    option = data.get("option", "")
    query = data.get("dataProduct", "")

    if option == "id":
        query = int(query)

    result = productDB.find_one({option: query}, {"_id": 0})
    return jsonify({"data": result, "query": query, "option": option})


@admin_product_bp.route("/update/updateProduct", methods=["POST"])
def update_product():
    data = request.get_json()
    info = data.get("info", "")
    id = data.get("id", "")
    productDB.update_one({"id": id}, {"$set": info})
    return jsonify({"info": info, "id": id})


